lib/ClimaCoreTempestRemap/src/onlineremap.jl

```diff
@@ -100,72 +160,110 @@ Generate the remapping weights from TempestRemap, returning a `LinearMap` object
+        overlap_mesh(meshfile_overlap, meshfile_source, meshfile_target)
+        remap_weights(
+            weightfile,
+            meshfile_source,
+            meshfile_target,
+            meshfile_overlap;
+            in_type = in_type,
+            in_np = Spaces.Quadratures.degrees_of_freedom(
+                source_space.quadrature_style,
+            ),
+            out_type = out_type,
+            out_np = Spaces.Quadratures.degrees_of_freedom(
+                target_space.quadrature_style,
+            ),
+        )
+
+        # read weight data
+        weights, col_indices, row_indices = NCDataset(weightfile, "r") do ds_wt
+            (Array(ds_wt["S"]), Array(ds_wt["col"]), Array(ds_wt["row"]))
+        end
+        # TempestRemap exports in CSR format (i.e. row_indices is sorted)
+
+        # TODO: add in extra rows to avoid DSS step
+        #  - for each unique node, we would add extra rows for all the duplicates, with the same column
+        #  - ideally keep in CSR format
+        #  e.g. iterate by row (i.e. target node):
+        #   - if new node, then append it
+        #   - if not new, copy previous entries
+        #  - requires a mechanism to query whether find the first common node of a given node
+
+
+        # we need to be able to look up the indices of unique nodes
+        source_unique_idxs =
+            in_type == "cgll" ? collect(Spaces.unique_nodes(source_space)) :
+            collect(Spaces.all_nodes(source_space))
+        target_unique_idxs =
+            out_type == "cgll" ? collect(Spaces.unique_nodes(target_space)) :
+            collect(Spaces.all_nodes(target_space))
 
-    # read weight data
-    weights, col_indices, row_indices = NCDataset(weightfile, "r") do ds_wt
-        (Array(ds_wt["S"]), Array(ds_wt["col"]), Array(ds_wt["row"]))
+        # re-order to avoid unnecessary allocations
+        source_unique_idxs_i =
+            map(col -> source_unique_idxs[col][1][1], col_indices)
+        source_unique_idxs_j =
+            map(col -> source_unique_idxs[col][1][2], col_indices)
+        source_unique_idxs_e =
+            map(col -> source_unique_idxs[col][2], col_indices)
+        target_unique_idxs_i =
+            map(row -> target_unique_idxs[row][1][1], row_indices)
+        target_unique_idxs_j =
+            map(row -> target_unique_idxs[row][1][2], row_indices)
+        target_unique_idxs_e =
+            map(row -> target_unique_idxs[row][2], row_indices)
+
+        source_unique_idxs =
+            (source_unique_idxs_i, source_unique_idxs_j, source_unique_idxs_e)
+        target_unique_idxs =
+            (target_unique_idxs_i, target_unique_idxs_j, target_unique_idxs_e)
+    else
+        weights = nothing
+        source_unique_idxs = nothing
+        target_unique_idxs = nothing
+        col_indices = nothing
+        row_indices = nothing
+    end
+
+    if !(comms_ctx isa ClimaComms.SingletonCommsContext)
+        root_pid = 0
```

Comment:
Would a barrier make sense here? I guess this is only called during init, so there shouldn't be a huge performance hit.

Reply:
Yeah, I think this is a good idea to make sure all processes have the correct information before creating the `LinearMap`.